# Import Libraries

In [1]:
import pandas as pd

# Loading Dataset

In [2]:
df = pd.read_csv('ca-oshpd-gachospital-hospitalbuildingdata-03092023.csv')

# Initial Inspection

In [3]:
df.head()

County Code  Perm ID     Facility Name     City Building Nbr  \
0  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01278   
1  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01279   
2  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01280   
3  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01281   
4  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01282   

        Building Name                 Building Status SPC Rating *  \
0   Original Hospital  No Gen Acute Care - OSHPD Bldg          NaN   
1       Stephens Wing                      In Service            2   
2           West Wing                      In Service            2   
3          South Wing                      In Service           3s   
4  Radiology Addition                      In Service           5s   

                                        Building URL  Height (ft)  Stories  \
0  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...        44.17      4.0   
1  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...        35.00      3.0   
2  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...          NaN      2.0   
3  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...          NaN      3.0   
4  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...          NaN      2.0   

                         Building Code  Building Code Year  Year Completed  \
0                              Unknown                 NaN          1926.0   
1     1952 Uniform Building Code (UBC)              1952.0          1956.0   
2     1964 Uniform Building Code (UBC)              1964.0          1968.0   
3  1976 California Building Code (CBC)              1976.0          1983.0   
4  1985 California Building Code (CBC)              1985.0          1995.0   

                                      AB 1882 Notice   Latitude   Longitude  \
0                                                NaN  37.762657 -122.253899   
1  This building does not significantly jeopardiz...  37.762657 -122.253899   
2  This building does not significantly jeopardiz...  37.762657 -122.253899   
3                                                NaN  37.762657 -122.253899   
4                                                NaN  37.762657 -122.253899   

   Count  
0      1  
1      1  
2      1  
3      1  
4      1

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3932 entries, 0 to 3931
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   County Code         3932 non-null   object 
 1   Perm ID             3932 non-null   int64  
 2   Facility Name       3932 non-null   object 
 3   City                3932 non-null   object 
 4   Building Nbr        3932 non-null   object 
 5   Building Name       3932 non-null   object 
 6   Building Status     3932 non-null   object 
 7   SPC Rating *        3337 non-null   object 
 8   Building URL        3932 non-null   object 
 9   Height (ft)         2020 non-null   float64
 10  Stories             3302 non-null   float64
 11  Building Code       3932 non-null   object 
 12  Building Code Year  3856 non-null   float64
 13  Year Completed      3550 non-null   float64
 14  AB 1882 Notice      709 non-null    object 
 15  Latitude            3932 non-null   float64
 16  Longit

# 1. Inspecting the data tidiness

During this stage, it's important to know if:
* Every column is a variable
* Every row is an observation
* Every cell is a single value

## 1.1 Investigating the `Building Code` column

In [8]:
df['Building Code'].value_counts()

Building Code
2001 California Building Code (CBC)    327
1973 California Building Code (CBC)    292
1979 California Building Code (CBC)    279
1985 California Building Code (CBC)    241
1989 California Building Code (CBC)    228
                                      ... 
1961 County of Los Angeles (LAC)         1
1959 County of Los Angeles (LAC)         1
1964 County of Los Angeles (LAC)         1
1966 County of Los Angeles (LAC)         1
2022 California Building Code (CBC)      1
Name: count, Length: 67, dtype: int64

In [9]:
df['Building Code'].describe()

count                                    3932
unique                                     67
top       2001 California Building Code (CBC)
freq                                      327
Name: Building Code, dtype: object

It looks like the bulding code as well as the year are both stored in the column, these can be split into two columns: `building_code` and `year`

## 1.2 Investigating unnecessary values

In [10]:
df.describe()

Perm ID  Height (ft)      Stories  Building Code Year  \
count   3932.000000  2020.000000  3302.000000         3856.000000   
mean   12157.352238    29.460847     1.958207         1985.519710   
std     2153.781276    27.212350     1.796883           19.356151   
min    10006.000000     0.000000     0.000000         1927.000000   
25%    10677.000000    13.000000     1.000000         1973.000000   
50%    11510.000000    17.000000     1.000000         1985.000000   
75%    12662.000000    36.000000     2.000000         2001.000000   
max    18232.000000   195.500000    15.000000         2022.000000   

       Year Completed     Latitude    Longitude   Count  
count     3550.000000  3932.000000  3932.000000  3932.0  
mean      1988.371549    35.670429  -119.428604     1.0  
std         20.356176     2.257315     2.045018     0.0  
min       1902.000000    32.618900  -124.194000     1.0  
25%       1973.000000    33.906395  -121.452156     1.0  
50%       1990.000000    34.325735  -118.463211     1.0  
75%       2005.000000    37.716740  -117.871239     1.0  
max       2022.000000    41.774509  -114.595116     1.0

The `Count` column does not add any useful information to the dataset.

## 1.3 Investigating different observational units

It can be see that there are lots of duplicate values, especially in the `County Code`, `Perm ID`, `Facility Name`, `City` and others. This can be visually seen when using the `.head()` function

In [11]:
df.head()

County Code  Perm ID     Facility Name     City Building Nbr  \
0  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01278   
1  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01279   
2  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01280   
3  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01281   
4  01 - Alameda    11210  Alameda Hospital  Alameda    BLD-01282   

        Building Name                 Building Status SPC Rating *  \
0   Original Hospital  No Gen Acute Care - OSHPD Bldg          NaN   
1       Stephens Wing                      In Service            2   
2           West Wing                      In Service            2   
3          South Wing                      In Service           3s   
4  Radiology Addition                      In Service           5s   

                                        Building URL  Height (ft)  Stories  \
0  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...        44.17      4.0   
1  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...        35.00      3.0   
2  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...          NaN      2.0   
3  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...          NaN      3.0   
4  https://esp.oshpd.ca.gov/CitizenAccess/Cap/Cap...          NaN      2.0   

                         Building Code  Building Code Year  Year Completed  \
0                              Unknown                 NaN          1926.0   
1     1952 Uniform Building Code (UBC)              1952.0          1956.0   
2     1964 Uniform Building Code (UBC)              1964.0          1968.0   
3  1976 California Building Code (CBC)              1976.0          1983.0   
4  1985 California Building Code (CBC)              1985.0          1995.0   

                                      AB 1882 Notice   Latitude   Longitude  \
0                                                NaN  37.762657 -122.253899   
1  This building does not significantly jeopardiz...  37.762657 -122.253899   
2  This building does not significantly jeopardiz...  37.762657 -122.253899   
3                                                NaN  37.762657 -122.253899   
4                                                NaN  37.762657 -122.253899   

   Count  
0      1  
1      1  
2      1  
3      1  
4      1

and confirmed using the `.unique()`  and `.nunique` functions

In [18]:
len(df['Facility Name'].unique())

415

In [19]:
df.nunique()

County Code             56
Perm ID                425
Facility Name          415
City                   254
Building Nbr          3932
Building Name         2708
Building Status         14
SPC Rating *            12
Building URL          3932
Height (ft)            694
Stories                 15
Building Code           67
Building Code Year      43
Year Completed         101
AB 1882 Notice           2
Latitude               424
Longitude              424
Count                    1
dtype: int64

415 unique facility names for around 4000 rows. A new table can be created containing the main hospital location information containing the columns with lots of repeated values (like the columns mention previously as well as others, which contain basic locational information about the hospital) and another table can created containing the rest of the information that are specifically related to a certain hospital building, creating a one to many relationship between a hospital table and its buildings.